internal/iamauth/token.go

```diff
@@ -0,0 +1,335 @@
+
+type IAMEntityType string
+
+const (
+	IAMEntityTypeNone IAMEntityType = ""
+	IAMEntityTypeRole               = "role"
+	IAMEntityTypeUser               = "user"
+)
+
+func NewBearerToken(loginToken string, config *Config) (*BearerToken, error) {
+	token := &BearerToken{config: config}
+	if err := json.Unmarshal([]byte(loginToken), &token); err != nil {
+		return nil, err
+	}
+
+	if err := token.validate(); err != nil {
+		return nil, err
+	}
+
+	if config.EnableIAMEntityDetails {
+		method, err := token.getHeader(token.config.GetEntityMethodHeader)
+		if err != nil {
+			return nil, err
+		}
+
+		url, err := token.getHeader(token.config.GetEntityURLHeader)
+		if err != nil {
+			return nil, err
+		}
+
+		headerJson, err := token.getHeader(token.config.GetEntityHeadersHeader)
+		if err != nil {
+			return nil, err
+		}
+
+		var header http.Header
+		if err := json.Unmarshal([]byte(headerJson), &header); err != nil {
+			return nil, err
+		}
+
+		body, err := token.getHeader(token.config.GetEntityBodyHeader)
+		if err != nil {
+			return nil, err
+		}
+
+		parsedUrl, err := token.parsedCallerIdentityURL.Parse(url)
+		if err != nil {
+			return nil, err
+		}
+
+		token.getIAMEntityMethod = method
+		token.getIAMEntityBody = body
+		token.getIAMEntityURL = url
+		token.getIAMEntityHeader = header
+		token.parsedIAMEntityURL = parsedUrl
+
+		entityType, err := token.validateIAMEntityBody()
+		if err != nil {
+			return nil, err
+		}
+		token.iamEntityType = entityType
+
+		// TODO:
+		// if err := t.validateServerIDHeader(); err != nil {
+		// 	return err
+		// }
+		// if err := t.validateAllowedSTSHeaderValues(); err != nil {
+		// 	return err
+		// }
+
+	}
+	return token, nil
+}
+
+// https://github.com/hashicorp/vault/blob/b17e3256dde937a6248c9a2fa56206aac93d07de/builtin/credential/aws/path_login.go#L1178
+func (t *BearerToken) validate() error {
+	if t.getCallerIdentityMethod != "POST" {
+		return fmt.Errorf("iam_http_request_method must be POST")
+	}
+	if err := t.validateGetCallerIdentityBody(); err != nil {
+		return err
+	}
+	if err := t.validateAllowedSTSHeaderValues(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// https://github.com/hashicorp/vault/blob/b17e3256dde937a6248c9a2fa56206aac93d07de/builtin/credential/aws/path_login.go#L1439
+func (t *BearerToken) validateGetCallerIdentityBody() error {
+	qs, err := url.ParseQuery(t.getCallerIdentityBody)
+	if err != nil {
+		return err
+	}
+	for k, v := range qs {
+		switch k {
+		case "Action":
+			if len(v) != 1 || v[0] != "GetCallerIdentity" {
+				return fmt.Errorf("iam_request_body must have 'Action=GetCallerIdentity'")
+			}
+		case "Version":
+		// Will assume for now that future versions don't change
+		// the semantics
+		default:
+			// Not expecting any other values
+			return fmt.Errorf("iam_request_body contains unexpected values")
+		}
+	}
+	return nil
+}
+
+func (t *BearerToken) validateIAMEntityBody() (IAMEntityType, error) {
```

Comment:
TODO: refactor into common response body parsing/validating function(s) (and avoid the loop label - ew)